Add docs for asset caching x-script usage (#61)
microsoft/vcpkg-docs: vcpkg/examples/assetcaching-script.md

---

> title: Asset Caching with an alternate storage source

Reviewer: I think this example should focus more on being a "How To" or a "Tutorial". The objective of a How To article is to first promise the user where they will end up (with an `x-script` asset cache pointing at NuGet in ADO) and then successfully transport them there (via concrete steps: "First, do X." "Then, do Y.").

(consider browsing https://diataxis.fr/how-to-guides/ for more philosophical inspiration)

```suggestion
title: How to create an x-script Asset Caching source for NuGet
```

Reviewer: If you change the title to reflect it being a how-to, you probably should also change the file name.

description: Learn about adding a new storage source
ms.date: 4/18/2023
---

# Asset Caching with an alternate storage source

[!INCLUDE [experimental](../../includes/experimental.md)]

> Your scenario might need a different storage source such as NuGet feed, Universal Packages, or something else.

Reviewer: As an continuation of the above point, a How-To shouldn't distract the user with other possibilities.

In this example, we will set up a NuGet feed as a source, meaning we will use `nuget.exe` to pull and push assets.

First, we will set up the environment before any call to `vcpkg install`:
```powershell

> $env:X_VCPKG_ASSET_SOURCES="clear;x-script,C:/path/to/sources.bat {url} {sha512} {dst};x-block-origin"

Reviewer: I think `sources.bat` is a bit of a confusing name; maybe `assetprovider.bat`?

$env:NUGET=C:/path/to/nuget.exe
$env:VCPKG_KEEP_ENV_VARS=NUGET
```

> `X_VCPKG_ASSET_SOURCES` passes in a script `sources.bat` which will handle `nuget` push and pull commands. The `vcpkg` environment needs to use a copy of `nuget.exe` which can be supplied by the environment with [VCPKG_KEEP_ENV_VARS](config-environment.md#vcpkg_keep_env_vars).

Reviewer: ```suggestion
[`X_VCPKG_ASSET_SOURCES`](../users/config-environment.md#x_vcpkg_asset_sources) defines a single [`x-script`](../users/assetcaching.md#x-script) asset source while blocking fallback to the original url ([`x-block-origin`](../users/assetcaching.md#x-block-origin)). The `x-script` source is defined to call a script `sources.bat` which will handle `nuget` push and pull commands. This script must be able to run both inside and outside the inner build environment, so we pass it the path to `nuget.exe` in the environment with `NUGET` and [`VCPKG_KEEP_ENV_VARS`](config-environment.md#vcpkg_keep_env_vars).
```

The script `sources.bat` checks if the asset is stored in the feed, and if so, installs it to the destination path. If the asset is not stored on the feed, the script fetches the url via curl, packs the asset in a `nuget` package and pushes it to the feed.

> `sources.bat`

Reviewer: As a point of style, we should not introduce files by simply having their name in backticks and then the content -- instead, we should use prose to instruct the user to create a file name so-and-so.

(example from the dotnet docs: https://learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?view=aspnetcore-7.0&tabs=visual-studio#the-model-and-database-context-classes)

```suggestion
Create a `sources.bat` file in your project with the following content:
```

This should probably be reordered above the environment setting steps above

```bat
set url=%1
set sha512=%2
set dst=%3
set "_dst=%dst:/=\%"
set "_sha512=%sha512:~0,90%"

> cd %dst%/../

Reviewer: ```suggestion
cd /d %~dp3
```
Doesn't need to resolve `/../` and successfully changes drives!

%NUGET% install %sha512:~0,90% -Source <feed url>
if exist %_sha512%.1.0.0 (
    echo "Pull from the NuGet feed"
    cd %_sha512%.1.0.0

    REM Assume both are files not directories
    echo "F" | xcopy /f *.part %_dst%
) else (
    echo "Fetch from the url"
    curl.exe -L %url% --create-dirs --output %dst%

>     cd %dst%/../

Reviewer: You already `cd`'d above.

    mkdir tempdir
    cd tempdir
    xcopy /f C:\path\to\test.nuspec .
    echo %_dst%
    xcopy /f %_dst% .

>     %NUGET% pack test.nuspec -Properties sha=%_sha512%
>     %NUGET% push %_sha512%.1.0.0.nupkg -Source <feed url>

Reviewer: ```suggestion
    %NUGET% pack C:\path\to\test.nuspec -BasePath %~dp3 -Properties "sha=%_sha512%;file=%dst%" -OutputDirectory %TEMP%
    %NUGET% push %TEMP%\%_sha512%.1.0.0.nupkg -Source <feed url>
```

)
```

> `test.nuspec`

Reviewer: ```suggestion
`asset-source.nuspec`
```
